src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,150 @@
+					: undefined;
+
+			const inflightRequest = new Request(request, {
+				signal: abortController ? abortController.signal : undefined
+			});
+
+			this._inflightRequests[key] = {
+				inflightRequest,
+				abortController,
+				dedupedRequests: [],
+				nextReqId: 0
+			};
+
+			newRequest = true;
 		}
 
-		const result = next(request);
-		if (result && result instanceof Promise) {
-			this._inflightRequests[key] = { count: 1 };
-			this._inflightRequests[key].action = result
-				.then(function(response) {
-					const usedMultiple = this._inflightRequests[key].count !== 1;
-					delete this._inflightRequests[key];
-					return this._clone(response, usedMultiple);
-				}.bind(this))
-				.catch(function(err) {
-					delete this._inflightRequests[key];
-					return Promise.reject(err);
-				}.bind(this));
-
-			return this._inflightRequests[key].action;
+		const dedupedRequest = new Promise((resolve, reject) => {
+			const inflightRequestEntry = this._inflightRequests[key];
+			const reqId = inflightRequestEntry.nextReqId++;
+			/** @type {DeduplicatedRequest} */
+			const newDedupedRequest = {
+				resolvers: { resolve, reject },
+				reqId
+			};
+
+			inflightRequestEntry.dedupedRequests.push(newDedupedRequest);
+
+			if (request.signal) {
+				/**
+				 * @param {AbortSignal} signal
+				 * @param {string} key
+				 * @param {number} reqId
+				 */
+				const addAbortEventListener = (signal, key, reqId) => {
+					const handler = () => {
+						signal.removeEventListener('abort', handler);
+
+						const inflightRequestEntry = this._inflightRequests[key];
+						if (!inflightRequestEntry) {
+							return;
+						}
+
+						const dedupedReqIndex = inflightRequestEntry.dedupedRequests.findIndex(r => r.reqId === reqId);
+						if (dedupedReqIndex < 0) {
+							return;
+						}
+
+						inflightRequestEntry.dedupedRequests.splice(dedupedReqIndex, 1);
+
+						if (inflightRequestEntry.dedupedRequests.length === 0 && inflightRequestEntry.abortController) {
+							inflightRequestEntry.abortController.abort();
+						}
+						reject(new DOMException('Request was aborted.', 'AbortError'));
+					};
+
+					newDedupedRequest.removeAbortListener = () => signal.removeEventListener('abort', handler);
+					signal.addEventListener('abort', handler);
+				};
+
+				addAbortEventListener(request.signal, key, reqId);
+			}
+		});
+
+		if (!newRequest) {
+			return dedupedRequest;
 		}
 
-		return result;
+		next(this._inflightRequests[key].inflightRequest)
+			.then((response) => {
+				const dedupedRequests = this._inflightRequests[key].dedupedRequests;
+				delete this._inflightRequests[key];
+
+				const res = dedupedRequests.length > 1
+					? this._clone(response)
+					: response;
+
+				for (const dedupedRequest of dedupedRequests) {
+					if (typeof dedupedRequest.removeAbortListener === 'function') {
+						dedupedRequest.removeAbortListener();
+					}
+					dedupedRequest.resolvers.resolve(res);
+				}
+			})
+			.catch((err) => {
```

Comment:
Doesn't look like this `catch` handler is expecting to handle errors from the above `then`
```suggestion
			}, (err) => {
```